File: pydbe/econ/cashflow.py
from pydbe._core import rs_cal_pv, rs_cal_pv_from_cf, rs_pv_unispread


def calculate_pv(cashflow: float, discount_rate: float, timestep: float) -> float:
    """
    Calculate the present value of cashflow.

    Parameters
    ----------
    cashflow : float
        The cashflow to be discounted.
    discount_rate : float
        The discount rate to be applied to the cashflow.
    timestep : float
        The future time step at which the cashflow occurs.

    Returns
    -------
    float
        The present value of the cashflow.
    """
    return rs_cal_pv(cashflow, discount_rate, timestep)


def calculate_pv_from_cashflows(cashflows: list[float], discount_rate: float) -> float:
    """
    Calculate the present value of a series of cashflows.

    Parameters
    ----------
    cashflows : list[float]
        A list of cashflows to be discounted.
    discount_rate : float
        The discount rate to be applied to the cashflows.

    Returns
    -------
    float
        The present value of the cashflows.
    """
    return rs_cal_pv_from_cf(cashflows, discount_rate)


def calculate_pv_unispread(
    t_pv: float, t_steps: int, discount_rate: float, w_init: bool = False
) -> float:
    """
    Calculate the uniform nominal value that will need to be pay out at each time step
    to achieve the targeted present value at the present time step.

    Parameters
    ----------
    t_pv : float
        The targeted present value at the present time step.
    t_steps : int
        The number of time steps over which the cashflow will be distributed.
    rate : float
        The discount rate to be applied to the cashflow.
    w_init : bool, optional
        Whether to include step 0 in the calculation. The default is False.

    Returns
    -------
    float
        The uniform nominal value that will need to be pay out at each time step to
        achieve the targeted present value at the present time step.
    """
    return rs_pv_unispread(t_pv, t_steps, discount_rate, w_init)

File: crates/core/src/econ/cashflow.rs
// src/econ/cashflow.rs
#[inline]
pub fn cal_pv(val: f64, rate: f64, step: f64) -> f64 {
    val / (1.0 + rate).powf(step)
}

pub fn cal_pv_from_cf(cf: &[f64], rate: f64) -> f64 {
    cf.iter()
        .enumerate()
        .map(|(i, &val)| cal_pv(val, rate, i as f64))
        .sum()
}

/// Spreading a target PV across selected future steps.
///
/// This function takes a target PV and returns the uniform nominal value that needs to
/// be achieved at each timestep in order to aggregate to the target PV.
///
/// # Arguments
/// * `t_pv` - The target present value to be achieved
/// * `t_steps` - Number of steps (excluding step 0) into the future in which the target
///               is expected to be achieved
/// * `rate` - The discount rate against which the cashflow is evaluated at
/// * `w_init` - Whether to account for an undiscounted year 0 value
pub fn pv_unispread(t_pv: f64, t_steps: i32, rate: f64, w_init: bool) -> f64 {
    let mut start_idx: i32 = 0;
    if !w_init {
        start_idx = 1;
    }

    let fact: f64 = (start_idx..=t_steps)
        .map(|t_step| (1.0 + rate).powf(t_step as f64))
        .sum();
    t_pv / fact
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cal_pv_with_std_inputs_return_expected_val() {
        let out = cal_pv(100.00, 0.12, 1.0);
        assert!(
            (out - 89.2857).abs() < 0.1,
            "Result: {out}, Expected: 89.2857",
        );
    }

    #[test]
    fn test_cal_pv_from_cf_with_std_inputs_return_expected_val() {
        let _in: [f64; 2] = [100.00, 100.00];
        let out = cal_pv_from_cf(&_in, 0.12);
        assert!(
            (out - 189.2857).abs() < 0.1,
            "Result: {out}, Expected: 189.2857",
        );
    }

    #[test]
    fn test_pv_unispread_with_init_returns_expected_val() {
        let out = pv_unispread(100.00, 2, 0.12, true);
        assert!(
            (out - 29.6348).abs() < 0.1,
            "Result: {out}, Expected: 29.6348",
        );
    }

    #[test]
    fn test_pv_unispread_without_init_returns_expected_val() {
        let out = pv_unispread(100.00, 2, 0.12, false);
        assert!(
            (out - 42.1159).abs() < 0.1,
            "Result: {out}, Expected: 42.1159",
        );
    }
}
